src/greq/msg.rs

```diff
@@ -0,0 +1,438 @@
+    /// It will be decrypted using AES-256 GCM and part of the message header will be used as
+    /// additional authenticated data (AAD).
+    ///
+    /// # Arguments
+    ///
+    /// * `msg_type`: Type of the command stored in the message payload
+    /// * `msg_seqno`: VMPL0 sequence number that was used in the message.
+    /// * `vmpck0`: VMPCK0 key, it will be used to decrypt the message
+    /// * `outbuf`: buffer that will be used to store the decrypted message payload
+    ///
+    /// # Returns
+    ///
+    /// * Success
+    ///     * usize: Number of bytes written to `outbuf`
+    /// * Error
+    ///     * [`SvsmReqError`]
+    pub fn decrypt_get(
+        &mut self,
+        msg_type: SnpGuestRequestMsgType,
+        msg_seqno: u64,
+        vmpck0: &[u8; VMPCK_SIZE],
+        outbuf: &mut [u8],
+    ) -> Result<usize, SvsmReqError> {
+        self.hdr.validate(msg_type, msg_seqno)?;
+
+        let iv: [u8; IV_SIZE] = build_iv(msg_seqno);
+        let aad: &[u8] = self.hdr.get_aad_slice();
+
+        // In the Aes256Gcm decrypt API, the authtag must be provided postfix in the inbuf
+        let ciphertext_end = usize::from(self.hdr.msg_sz);
+        let tag_end: usize = ciphertext_end + AUTHTAG_SIZE;
+
+        // The message payload must be large enough to hold the ciphertext and
+        // the authentication tag.
+        //
+        // The outbuf must be large enough to hold the plaintext, which should
+        // have the same size as the ciphertext.
+        if tag_end > self.pld.len() || ciphertext_end > outbuf.len() {
+            return Err(SvsmReqError::invalid_parameter());
+        }
+
+        self.pld[ciphertext_end..tag_end].copy_from_slice(&self.hdr.authtag[..AUTHTAG_SIZE]);
+
+        // Payload with postfixed authtag
+        let inbuf = &self.pld[..tag_end];
+
+        let outbuf_len: usize = Aes256Gcm::decrypt(&iv, vmpck0, aad, inbuf, outbuf)?;
+
+        Ok(outbuf_len)
+    }
+}
+
+/// Build the initialization vector for AES-256 GCM
+fn build_iv(msg_seqno: u64) -> [u8; IV_SIZE] {
+    const U64_SIZE: usize = size_of::<u64>();
+    let mut iv = [0u8; IV_SIZE];
+
+    iv[..U64_SIZE].copy_from_slice(&msg_seqno.to_ne_bytes());
+    iv
+}
+
+/// Data page(s) the hypervisor will use to store certificate data in
+/// an extended `SNP_GUEST_REQUEST`
+#[derive(Debug)]
+pub struct SnpGuestRequestExtData {
+    /// According to the GHCB spec, the data page(s) must be contiguous pages if
+    /// supplying more than one page and all certificate pages must be
+    /// assigned to the hypervisor (shared).
+    data: [u8; SNP_GUEST_REQ_MAX_DATA_SIZE],
+}
+
+impl SnpGuestRequestExtData {
+    /// Clear the first `n` bytes from data
+    pub fn nclear(&mut self, n: usize) -> Result<(), SvsmReqError> {
+        if n > self.data.len() {
+            return Err(SvsmReqError::invalid_parameter());
+        }
+        self.data[..n].fill(0);
```

Comment:
```suggestion
        self.data
            .get_mut(..n)
            .ok_or_else(SvsmReqError::invalid_parameter)?
            .fill(0);
```
Also applies to `copy_to_slice` and `is_nclear`.